### PR TITLE
Account memory for OrcDeletedRows

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java [hunk 1]
```diff
@@ -244,6 +251,10 @@ private RowId getRowId(int position)
                                 long row = BIGINT.getLong(page.getBlock(ROW_ID_INDEX), i);
                                 RowId rowId = new RowId(originalTransaction, bucket, statement, row);
                                 deletedRowsBuilder.add(rowId);
+                                rowCount++;
+                                if (rowCount % 1000 == 0) {
+                                    systemMemoryUsage.setBytes(rowCount * RowId.INSTANCE_SIZE);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:256]: `sizeOfObjectArray(rowCount) + rowCount * RowId.INSTANCE_SIZE`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java [hunk 2]
```diff
@@ -262,7 +273,9 @@ private RowId getRowId(int position)
                 throw new TrinoException(HIVE_CURSOR_ERROR, "Failed to read ORC delete delta file: " + path, e);
             }
         }
+
         deletedRows = deletedRowsBuilder.build();
+        systemMemoryUsage.setBytes((long) deletedRows.size() * RowId.INSTANCE_SIZE);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:278]: cast `(long)` above too

**Review by findepi — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java (hunk 1 above)
Comment (alexjo2144) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:256]: Is it worth doing setting this as we loop vs just doing it once at the end?
Also, can we safely ignore the return value of `setBytes`?

Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:256]: The benefit of doing that in the loop (I think) is that we can pause some other query for a while, allowing us to get to the end of loop without OOM. Other than that it does not change much I think.
@findepi am I missing something?

**Comment by losipiuk** [comment]
AC

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java]: we should assume that building `deletedRows` list can exhaust node memory.
if we set memory usage only at the end, we still do not protect ourselves against node OOM.

> Also, can we safely ignore the return value of setBytes?

we cannot block here (can we?).
we could use `trySetBytes` here, perhaps.

**Review by findepi — APPROVED** [review]

**Comment by losipiuk** [comment]
> we cannot block here (can we?).
It does not look like. Maybe with a significant refactoring.

`trySetBytes` I was thinking about that. But then what if it returns that we cannot allocate. Throw exception ourselves?

**Comment by findepi** [comment]
> > we could use trySetBytes here, perhaps.
> 
> I was thinking about that. But then what if it returns that we cannot allocate. Throw exception ourselves?

yes, we would throw then.

cc @sopel39 for `LocalMemoryContext.setBytes` vs `trySetBytes`.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java [hunk 3]
```diff
@@ -244,6 +252,10 @@ private RowId getRowId(int position)
                                 long row = BIGINT.getLong(page.getBlock(ROW_ID_INDEX), i);
                                 RowId rowId = new RowId(originalTransaction, bucket, statement, row);
                                 deletedRowsBuilder.add(rowId);
+                                rowCount++;
+                                if (rowCount % 1000 == 0) {
```
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:256]: add a comment why

**Review by sopel39 — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java (hunk 3 above)
Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java]: dropped as data is not passed to operator context immediately.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java [hunk 4]
```diff
@@ -262,10 +269,19 @@ private RowId getRowId(int position)
                 throw new TrinoException(HIVE_CURSOR_ERROR, "Failed to read ORC delete delta file: " + path, e);
             }
         }
+
         deletedRows = deletedRowsBuilder.build();
+        // not updating memory usage in the loop, as deletedRows are built,  as recorded information is propagated
```
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:274]: nit: extra space before `as`

**Pull request merged**